torch/_functorch/aot_autograd.py

```diff
@@ -3357,6 +3655,264 @@ def forward(*runtime_args):
 
     return forward
 
+"""
+This function takes in a module, and returns:
+(1) an FX graph that can be exported
+(2) some metadata about the graph
+
+If `trace_joint=True` we will return a joint graph of the forward + backward.
+
+The traced FX graph will have the following properties compared to the original module:
+(1) Inputs and outputs to the module will be pytree-flattened
+(2) Parameters and buffers on the module will be lifted into graph inputs,
+    graph_inputs = (*parameters, *buffers, *user_inputs)
+(3) The graph will be fully functionalized
+(4) Any input mutations will be converted into additional outputs in the graph,
+    meaning whoever calls this graph is responsible for applying the mutations
+    back to the original inputs.
+(5) If is_backward is provided the graph will return parameter gradients in addition to user outputs.
+    The graph output will look like:
+    graph_outputs = (*updated_inputs, *user_outputs, *param_gradients)
+
+There are also several restrictions on what modules can use this API. In particular:
+(1) If trace_joint is specified, we expect the loss function to be **fused**
+    into the module forward. One of the outputs to the forward must be a scalar loss,
+    which is specified with `output_loss_index`.
+    All other outputs to the forward are presumed to not require gradients.
+(2) This API cannot capture optimizers (although in theory we could build an API for this).
+(3) Metadata mutations on params/buffers/inputs are banned.
+(4) Data mutations on anything that requires gradients are banned (parameters)
+(5) If an input is mutated, it is not allowed to alias any other inputs.
+(6) Parameters must not be duplicated.
+"""
+def aot_export_module(
+    mod: nn.Module,
+    args,
+    *,
+    decompositions: Optional[Dict] = None,
+    # If true, we'll return a joint forward-backward graph,
+    # As well as metadata on the loss + gradients in the backward.
+    trace_joint: bool,
+    # If trace_joint is True, we expect your module to return a scalar loss.
+    # Your module can return multiple outputs, so you must specify which output the loss is.
+    output_loss_index: Optional[int] = None,
```

Comment:
If `trace_joint = True` but `output_loss_index is None`, what is the behavior?

Reply:
This should be invalid input.

Reply:
fixed